### ISSUE TITLE
Improve running time of the POST /api/analysis_batches/ endpoint

### ISSUE THREAD

**Issue by CloudNiner** [azavea/pfb-network-connectivity#578]
The AnalysisBatch create process involves two HTTP requests for each AnalysisJob created. One is to update the Neighborhood zipfile stored on S3 in `Neighborhood.set_boundary_file()`, the other is to make a Boto request to Batch to submit the job.

The combination of these two requests for hundreds of Neighborhoods in a batch shapefile results in a POST request that can take a significant chunk of time and grows linearly with the number of Neighborhoods such that 50+ Neighborhoods currently throw an error.

The endpoint should be refactored and/or the gunicorn+nginx+ALB timeouts should be increased to allow long running requests. A spinner on the AnalysisBatch create admin page while the request is pending would be a nice improvement as well.

**Comment by CloudNiner** [comment]
A WIP branch is pushed to https://github.com/azavea/pfb-network-connectivity/tree/feature/awf/sensibly-synchronous-batch-create.

This attempts to address the issue with a combination of running AnalysisBatch.submit() in a separate Thread, and increases HTTP timeouts from 60s -> 120s which gives enough time for the POST to complete successfully for the largest know shapefiles (which have 200 Neighborhoods in them).

**Comment by CloudNiner** [comment]
I should have realized this earlier, but we're constrained by the nature of Django + gunicorn limiting processes to the timeout window. Any async process run within the gunicorn process remains attached to that process, meaning it will still be killed when the timeout arrives. 

So our options appear to be two-fold:
1. Make request timeouts really high (3mins) and continue to run this code within the request/response cycle
2. Attempt to use subprocess and make a django management call in a detached child process. This _should_ work, but has some potential traps around calling to ensure it actually becomes detached. There is no way to determine success or failure, if that becomes useful
3. Add an async task queue and run the Batch.submit() on that. [django-q](https://django-q.readthedocs.io/en/latest/) seems like a decent, lightweight Celery alternative. We could use the Django ORM broker, and add an additional ECS task that runs the task workers, which is a Django management command. 

I'd rank preference from highest to lowest as 3 - 2 - 1. Bumping request timeout is bad for many reasons. Using subprocess isn't flexible at all, is prone to future breakage, and we won't know when it does break because we don't get any insight into the spawned process. There's also some measure of debugging time to figure out how to actually make the call such that it doesn't run within the calling process. While I'm concerned about an untested library, I don't see it taking more than the time allotted remaining on this card to add an ECS task, configure the task runner, and implement using the ORM broker, which should be fine for such light use. The additional resource requirements are low and should fit within the current ECS cluster. I didn't see any other async Django queue plugins that are as lightweight as this one, and maintained.

Thoughts @KlaasH?

**Comment by KlaasH** [comment]
Yeah, 1 and 2 seem to have unacceptable drawbacks.  Setting up a whole new service for this seems like a surprisingly heavy-weight solution to what seemed like a smallish problem, but if there's no viable alternatives then so it goes.  And `django-q` does sound pretty easy to work with.

We do already have an asynchronous processing setup, of course.  Is there a technical reason why it wouldn't work to add batch creation as a Batch job?  I haven't come up with any, but I'm also having trouble discerning how much work it would be and how reusable our existing Batch tooling would be if we wanted to make it a Batch job (that spawns and submits other Batch jobs).

**Comment by CloudNiner** [comment]
Ah. Yes, of course. I should have added Batch as an option in the discussion above. I excluded it for a few reasons:

We would have to add the infrastructure for an additional Batch job definition + queue, which by itself at least as much boilerplate as for the django-q strategy. The existing job definitions are [ad hoc in a python script](https://github.com/azavea/pfb-network-connectivity/blob/develop/deployment/aws-batch/update_job_definition.py) and aren't what I would call robust. Perhaps it would be a more compelling alternative if we had already swapped to the new AWS Batch terraform resources, but we haven't. 

Running in Batch would also require either autoscaling to be implemented so that the job can run in the existing cluster without manually adjusting instance counts, or run in a separate managed cluster using the existing Django container. Again, we haven't swapped to the AWS Batch terraform resources, making this a much larger lift because we'd have to provision a cluster outside of terraform in some way if we went down the latter path.

**Comment by CloudNiner** [comment]
I should note that if the Terraform upgrade and AWS Managed Batch compute environment tasks were completed, then implementing this as a batch job using the Django container and having that run a management command would likely be the correct implementation choice. The only downside in that case (versus django-q) would be a slightly slower start up time, because there's a bit of a delay for a EC2 instance to start up, download the container image and run the job. But I think a few minute delay in startup time would be worth consistency in the async processing infrastructure for a job that will see light, administrator-based usage.
